pkg/ccl/changefeedccl/tableset/watcher.go

```diff
@@ -0,0 +1,536 @@
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeedbase"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkeys"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
+	"github.com/cockroachdb/cockroach/pkg/util/span"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/logtags"
+	"golang.org/x/sync/errgroup"
+)
+
+// Filter is a filter for a tableset.
+type Filter struct {
+	// DatabaseID is the ID of the database to watch. It is mandatory.
+	DatabaseID descpb.ID
+	// IncludeTables and ExcludeTables are mutually exclusive.
+	// TODO(#147420): replace with the protobuf introduced in the pr for this issue.
+	IncludeTables map[string]struct{}
+	ExcludeTables map[string]struct{}
+}
+
+func (f Filter) String() string {
+	return fmt.Sprintf("Filter{database_id=%d, include_tables=%v, exclude_tables=%v}",
+		f.DatabaseID, slices.Collect(maps.Keys(f.IncludeTables)),
+		slices.Collect(maps.Keys(f.ExcludeTables)))
+}
+
+func (f Filter) includes(tableInfo Table) bool {
+	if tableInfo.ParentID != f.DatabaseID {
+		return false
+	}
+	if len(f.ExcludeTables) > 0 {
+		_, ok := f.ExcludeTables[tableInfo.Name]
+		return !ok
+	}
+	if len(f.IncludeTables) > 0 {
+		_, ok := f.IncludeTables[tableInfo.Name]
+		return ok
+	}
+	return true
+}
+
+type Table struct {
+	descpb.NameInfo
+	ID descpb.ID
+}
+
+func (t Table) size() int64 {
+	return int64(t.NameInfo.Size()) + 8
+}
+
+func (t Table) String() string {
+	return fmt.Sprintf("Table{parent_id=%d, parent_schema_id=%d, name=%s, id=%d}",
+		t.NameInfo.ParentID, t.NameInfo.ParentSchemaID, t.NameInfo.Name, t.ID)
+}
+
+type TableDiff struct {
+	Added   Table
+	Dropped Table
+	AsOf    hlc.Timestamp
+}
+
+func (d TableDiff) size() int64 {
+	return d.Added.size() + d.Dropped.size() + int64(d.AsOf.Size())
+}
+
+func compareTableDiffsByTS(a, b TableDiff) int {
+	return a.AsOf.Compare(b.AsOf)
+}
+
+type TableSet struct {
+	Tables []Table
+	AsOf   hlc.Timestamp
+}
+
+// Watcher watches a tableset and buffers table diffs. It will notify waiters
+// when the resolved timestamp is advanced.
+type Watcher struct {
+	filter Filter
+
+	id      int64
+	execCfg *sql.ExecutorConfig
+	mon     *mon.BytesMonitor
+	acc     mon.BoundAccount // set on Start
+	started bool
```

Comment:
maybe `started` should go in `mu` to protect against concurrent starts?

Reply:
ehhhhhhhh